package me.srrapero720.watervision.common.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.BoolArgumentType;
import com.mojang.brigadier.arguments.FloatArgumentType;
import com.mojang.brigadier.arguments.IntegerArgumentType;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import me.srrapero720.watervision.WaterVision;
import me.srrapero720.watervision.client.screens.VisionScreen;
import me.srrapero720.watervision.common.network.PlayVideoPacket;
import me.srrapero720.watervision.common.network.VisionNetwork;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.command.v2.ClientCommandManager;
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.Screen;
import me.srrapero720.watervision.WaterVisionClient;
import me.srrapero720.watervision.common.network.*;
import net.minecraft.commands.CommandSourceStack;
import net.minecraft.commands.Commands;
import net.minecraft.commands.arguments.EntityArgument;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;

public class VisionCommands {
    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
        dispatcher.register(Commands.literal("playvideo")
                .requires(source -> !source.isPlayer() || source.hasPermission(4))
                .then(Commands.argument("url", StringArgumentType.string())
                        .then(Commands.argument("target", EntityArgument.players())
                                .executes(VisionCommands::openVideoScreen)
                                .then(Commands.argument("volume", IntegerArgumentType.integer(0, 100))
                                        .executes(VisionCommands::openVideoScreen)
                                        .then(Commands.argument("speed", FloatArgumentType.floatArg(0.25f, 2.0f))
                                                .executes(VisionCommands::openVideoScreen)
                                                .then(Commands.argument("stretch_video", BoolArgumentType.bool())
                                                        .executes(VisionCommands::openVideoScreen)
                                                        .then(Commands.argument("game_fade_duration", FloatArgumentType.floatArg(0.0f, 100.0f))
                                                                .executes(VisionCommands::openVideoScreen)
                                                                .then(Commands.argument("video_fade_duration", FloatArgumentType.floatArg(0.0f, 100.0f))
                                                                        .executes(VisionCommands::openVideoScreen)
                                                                        .then(Commands.argument("allow_controls", BoolArgumentType.bool())
                                                                                .executes(VisionCommands::openVideoScreen)
                                                                                .then(Commands.argument("allow_exit", BoolArgumentType.bool())
                                                                                        .executes(VisionCommands::openVideoScreen)
                                                                                )
                                                                        )
                                                                )

                                                        )

                                                )
                                        )
                                )
                        )
                )
        );


        dispatcher.register(Commands.literal("stopvideo")
                .requires(source -> !source.isPlayer() || source.hasPermission(4))
                .then(Commands.argument("target", EntityArgument.players())
                        .executes(VisionCommands::stopVideo)
                )
        );
        dispatcher.register(Commands.literal("playoverlay")
                .requires(source -> !source.isPlayer() || source.hasPermission(4))
                .then(Commands.argument("url", StringArgumentType.string())
                        .then(Commands.argument("target", EntityArgument.players())
                                .executes(VisionCommands::openVideoOverlay)
                        )
                )
        );

        dispatcher.register(Commands.literal("stopoverlay")
                .requires(source -> !source.isPlayer() || source.hasPermission(4))
                .then(Commands.argument("target", EntityArgument.players())
                        .executes(VisionCommands::closeVideoOverlay)
                )
        );
    }

    @Environment(EnvType.CLIENT)
    public static void registerClient(final CommandDispatcher<FabricClientCommandSource> dispatcher) {
        dispatcher.register(ClientCommandManager.literal("playvideoclient")
                .requires(source -> source.getEntity() != null)
                .then(ClientCommandManager.argument("url", StringArgumentType.string())
                        .executes(VisionCommands::openVideoScreenClient)
                        .then(ClientCommandManager.argument("volume", IntegerArgumentType.integer(0, 100))
                                .executes(VisionCommands::openVideoScreenClient)
                                .then(ClientCommandManager.argument("speed", FloatArgumentType.floatArg(0.25f, 2.0f))
                                        .executes(VisionCommands::openVideoScreenClient)
                                        .then(ClientCommandManager.argument("stretch_video", BoolArgumentType.bool())
                                                .executes(VisionCommands::openVideoScreenClient)
                                        )
                                )
                        )
                )
        );

        dispatcher.register(ClientCommandManager.literal("playoverlayclient")
                .requires(source -> source.getEntity() != null)
                .then(ClientCommandManager.argument("url", StringArgumentType.string())
                        .executes(VisionCommands::openVideoOverlayClient)
                )
        );

        dispatcher.register(ClientCommandManager.literal("stopoverlayclient")
                .requires(source -> source.getEntity() != null)
                .executes(VisionCommands::closeVideoOverlayClient)
        );
    }

    @Environment(EnvType.CLIENT)
    private static int openVideoScreenClient(CommandContext<FabricClientCommandSource> context) throws CommandSyntaxException {
        try {
            final var url = StringArgumentType.getString(context, "url");
            final var volume = getIntOrDefaultClient(context, "volume", 100);
            final var speed = getFloatOrDefaultClient(context, "speed", 1.0f);
            final var stretchVideo = getBoolOrDefaultClient(context, "stretch_video", false);

            // THIS IS A WORKARROUND BECAUSE... AMM... FABRIC SUCKS
            new Thread(() -> {
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
                Minecraft.getInstance().execute(() -> {
                    WaterVisionClient.openScreen(url, volume, speed, stretchVideo, WaterVisionClient.DEF_GAME_FADE_DURATION, WaterVisionClient.DEF_VIDEO_FADE_DURATION, true, true);
                });
            }).start();
            return 0;
        } catch (final Exception e) {
            context.getSource().sendError(Component.literal("Failed to open video screen, see log for more details"));
            WaterVision.LOGGER.error("Failed to execute /playvideoclient command", e);
        }

        return 1;
    }

    @Environment(EnvType.CLIENT)
    private static int openVideoOverlayClient(CommandContext<FabricClientCommandSource> context) throws CommandSyntaxException {
        try {
            final var url = StringArgumentType.getString(context, "url");

            WaterVisionClient.openOverlay(url);
            return 0;
        } catch (Throwable e) {
            context.getSource().sendError(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playoverlay command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }
        return 1;
    }

    @Environment(EnvType.CLIENT)
    private static int closeVideoOverlayClient(CommandContext<FabricClientCommandSource> context) throws CommandSyntaxException {
        try {

            WaterVisionClient.closeOverlay();
            return 0;
        } catch (Exception e) {
            context.getSource().sendError(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playoverlay command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }
        return 1;
    }

    private static int openVideoScreen(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
        try {
            final var url = StringArgumentType.getString(context, "url");
            final var players = EntityArgument.getPlayers(context, "target");
            final var volume = getIntOrDefault(context, "volume", 100);
            final var speed = getFloatOrDefault(context, "speed", 1.0f);
            final var stretchVideo = getBoolOrDefault(context, "stretch_video", false);
            final var gameFadeDuration = getFloatOrDefault(context, "game_fade_duration", 20.0f);
            final var videoFadeDuration = getFloatOrDefault(context, "video_fade_duration", 20.0f);
            final var allowControls = getBoolOrDefault(context, "allow_controls", true);
            final var exit = getBoolOrDefault(context, "allow_exit", true);

            for (final var player: players) {
                WaterVision.LOGGER.info("Opening videoscreen for {}", player.getName().getString());
                VisionNetwork.sendTo(new PlayVideoPacket(url, volume, speed, stretchVideo, gameFadeDuration, videoFadeDuration, allowControls, exit), player);
            }

            return 0;
        } catch (final Throwable e) {
            context.getSource().sendFailure(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playvideo command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }

        return 1;
    }

    private static int openVideoOverlay(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
        try {
            final var url = StringArgumentType.getString(context, "url");
            final var players = EntityArgument.getPlayers(context, "target");

            for (final var player: players) {
                VisionNetwork.sendTo(new PlayVideoOverlayPacket(url), player);
            }

            return 0;
        } catch (Exception e) {
            context.getSource().sendFailure(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playoverlay command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }
        return 1;
    }

    private static int closeVideoOverlay(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
        try {
            final var players = EntityArgument.getPlayers(context, "target");

            for (final var player: players) {
                VisionNetwork.sendTo(new StopVideoOverlayPacket(), player);
            }

            return 0;
        } catch (Exception e) {
            context.getSource().sendFailure(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playoverlay command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }
        return 1;
    }

    private static int stopVideo(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
        try {
            final var players = EntityArgument.getPlayers(context, "target");

            for (final var player: players) {
                VisionNetwork.sendTo(new StopVideoPacket(), player);
            }

            return 0;
        } catch (Exception e) {
            context.getSource().sendFailure(Component.literal("Failed to open video screen, see console for more details"));
            WaterVision.LOGGER.error("Failed to execute /playoverlay command", e);

            if (e instanceof CommandSyntaxException) {
                throw e;
            }
        }
        return 1;
    }

    private static String getStringOrDefault(CommandContext<CommandSourceStack> context, String name, String def) {
        try {
            return StringArgumentType.getString(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static int getIntOrDefault(CommandContext<CommandSourceStack> context, String name, int def) {
        try {
            return IntegerArgumentType.getInteger(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static float getFloatOrDefault(CommandContext<CommandSourceStack> context, String name, float def) {
        try {
            return FloatArgumentType.getFloat(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static boolean getBoolOrDefault(CommandContext<CommandSourceStack> context, String name, boolean def) {
        try {
            return BoolArgumentType.getBool(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static String getStringOrDefaultClient(CommandContext<FabricClientCommandSource> context, String name, String def) {
        try {
            return StringArgumentType.getString(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static int getIntOrDefaultClient(CommandContext<FabricClientCommandSource> context, String name, int def) {
        try {
            return IntegerArgumentType.getInteger(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static float getFloatOrDefaultClient(CommandContext<FabricClientCommandSource> context, String name, float def) {
        try {
            return FloatArgumentType.getFloat(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }

    private static boolean getBoolOrDefaultClient(CommandContext<FabricClientCommandSource> context, String name, boolean def) {
        try {
            return BoolArgumentType.getBool(context, name);
        } catch (IllegalArgumentException ex) {
            return def;
        }
    }
}
